File: java/src/main/java/com/mycompany/pt/PaymentTracker.java
/*
 * |-------------------------------------------------
 * | Copyright © 2017 Colin But. All rights reserved.
 * |-------------------------------------------------
 */
package com.mycompany.pt;

import com.mycompany.pt.io.Command;
import com.mycompany.pt.io.PaymentTrackerReader;
import com.mycompany.pt.scheduling.PaymentsScheduler;
import com.mycompany.pt.util.UsagePrinter;
import lombok.extern.slf4j.Slf4j;

import java.util.Scanner;

@Slf4j
public class PaymentTracker {

    private static final String INITIAL_INPUT_FILE_NAME = "initial_input_file.txt";


    public static void main(String[] args) {

        PaymentTrackerReader.readInputFile("src/main/resources/" + INITIAL_INPUT_FILE_NAME);

        new PaymentsScheduler().schedulePaymentListings();

        if (args.length > 1) {
            UsagePrinter.printUsage();
            throw new IllegalArgumentException("Invalid number of arguments supplied");
        }

        if (args.length == 1) {
            String optionalInputFile = args[0];
            PaymentTrackerReader.readInputFile(optionalInputFile);
        }

        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.print("Enter Something: ");

            String input = scanner.nextLine();

            if (Command.Q_COMMAND.getCommand().equalsIgnoreCase(input) || Command.QUIT_COMMAND.getCommand().equalsIgnoreCase(input)) {
                System.out.println("Exiting!");
                break;
            }

            Parser.getInstance().invoke(input);
        }

        scanner.close();

        System.exit(1);
    }

}

File: kotlin/src/main/kotlin/payment/tracker/PaymentTracker.kt
package payment.tracker

import payment.tracker.io.PaymentTrackerReader
import payment.tracker.schedule.PaymentScheduler
import payment.tracker.util.UsagePrinter
import java.lang.IllegalArgumentException
import java.util.*
import kotlin.collections.HashMap

val paymentListing = PaymentListing(HashMap())
val parser = Parser(paymentListing)
val paymentScheduler = PaymentScheduler()

const val DEFAULT_INPUT_FILE = "src/main/resources/initial_input_file.txt"

fun main(args: Array<String>) {

    paymentScheduler.schedulePaymentListing()

    when {
        args.isEmpty() -> readDefaultInputFile()
        args.size == 1 -> readInputFileFromCommandLine(args)
        args.size > 1 -> showErrorAndExit()
    }

    val scanner = Scanner(System.`in`)

    while(true) {
        print("Enter something: ")
        val line : String = scanner.nextLine()
        if (line.equals("quit", true) || line.equals("q", true)) {
            print("Exiting!")
            break
        }
        parser.invoke(line)
    }

    scanner.close()
}

fun readDefaultInputFile() = PaymentTrackerReader.readInputFile(DEFAULT_INPUT_FILE)
        .forEach { line -> parser.invoke(line) }

fun readInputFileFromCommandLine(args: Array<String>) = PaymentTrackerReader.readInputFile(args[0])
        .forEach { line -> parser.invoke(line) }

fun showErrorAndExit() {
    UsagePrinter.printUsage()
    throw IllegalArgumentException("Invalid number of arguments")
}